api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirObservationDaoImpl.java

```diff
@@ -175,4 +220,53 @@ protected Obs deproxyResult(Obs result) {
 		obs.setConcept(deproxyObject(obs.getConcept()));
 		return obs;
 	}
+	
+	int getMax(SearchParameterMap theParams) {
+		return ((NumberParam) theParams.getParameters(FhirConstants.MAX_SEARCH_HANDLER).get(0).getParam()).getValue()
+		        .intValue();
+		
+	}
+	
+	protected List<String> handleRanking(List<Object[]> list, int max) {
+		List<String> results = new ArrayList<>();
+		int cur_rank = 0;
+		
+		for (int var = 0; var < list.size() && cur_rank < max; var++) {
+			cur_rank++;
+			results.add((String) list.get(var)[0]);
+			
+			while (var + 1 < list.size() && list.get(var + 1)[2].equals(list.get(var)[2])) {
+				results.add((String) list.get(var + 1)[0]);
+				var++;
+			}
+		}
+		return results;
+		
+	}
+	
+	protected void handleGrouping(HashMap<Concept, List<Object[]>> groups, List<Object[]> observations) {
+		for (Object[] obs : observations) {
+			
+			if (groups.containsKey(obs[1])) {
+				groups.get(obs[1]).add(obs);
+			} else {
+				List<Object[]> list = new ArrayList<>();
+				list.add(obs);
+				groups.put((Concept) obs[1], list);
+			}
+		}
+	}
+	
+	protected List<String> getLastNUuids(HashMap<Concept, List<Object[]>> groups, int max) {
```

Comment:
Hadn't really noticed this before, but we should favour the abstract type as an argument wherever possible (e.g. `Map<Concept, List<Object[]>>` instead of `HashMap<>`). This allows us to later switch to another Map implementation, if necessary.